### PR TITLE
Extend circuit node_sets with simulation node sets

### PR DESCRIPTION
**Pull request by joni-herttuainen**
Also
* Add `update` method to node sets
* restrict `libsonata>=0.1.24`

### CODE REVIEW

File: bluepysnap/node_sets.py [hunk 1]
```diff
@@ -89,6 +90,28 @@ def from_dict(cls, content):
...
+        """Access (a copy of) the node sets contents."""
+        return copy.deepcopy(self._content)
+
+    @property
+    def to_libsonata(self):
+        """Libsonata instance of the NodeSets."""
+        return self._instance
+
+    def update(self, node_sets):
+        """Update the contents of the node set."""
```
Comment (GianlucaFicarelli) [inline bluepysnap/node_sets.py:104]: It should be mentioned in the docstring that `update` returns a set with the names of overwritten node_sets.

File: bluepysnap/simulation.py [hunk 2]
```diff
@@ -125,8 +136,17 @@ def simulator(self):
...
-        return NodeSets.from_file(path) if path else NodeSets.from_dict({})
+        try:
+            path = self.circuit.to_libsonata.node_sets_path
+        except BluepySnapError:
+            path = ""
+
+        node_sets = NodeSets.from_file(path) if path else NodeSets.from_dict({})
+
+        if path := self.to_libsonata.node_sets_file:
+            _warn_on_overwritten_node_sets(node_sets.update(NodeSets.from_file(path)))
```
Comment (GianlucaFicarelli) [inline bluepysnap/simulation.py:147]: For clarity I would split this in 2 lines: 1 to update the node_sets, 2 to print the warning
```suggestion
            overwritten = node_sets.update(NodeSets.from_file(path))
            _warn_on_overwritten_node_sets(overwritten)
```

File: bluepysnap/simulation.py [hunk 3]
```diff
@@ -125,8 +136,17 @@ def simulator(self):
     @cached_property
     def node_sets(self):
         """Returns the NodeSets object bound to the simulation."""
-        path = self.to_libsonata.node_sets_file
-        return NodeSets.from_file(path) if path else NodeSets.from_dict({})
+        try:
+            path = self.circuit.to_libsonata.node_sets_path
+        except BluepySnapError:
```
Comment (GianlucaFicarelli) [inline bluepysnap/simulation.py:141]: I see that `self.circuit` might raise BluepySnapError if the circuit config file doesn't exist.
Is the simulation config considered valid even in this case?
If we need to keep the check, a comment with the type of expected error may be useful.

File: bluepysnap/simulation.py [hunk 4]
```diff
@@ -125,8 +136,17 @@ def simulator(self):
...
+            path = ""
+
+        node_sets = NodeSets.from_file(path) if path else NodeSets.from_dict({})
```
Comment (GianlucaFicarelli) [inline bluepysnap/simulation.py:144]: We could get a copy of `self.circuit.node_sets` instead, but it makes sense only if it makes the code simpler.

File: bluepysnap/simulation.py (hunk 2 above)
Comment (joni-herttuainen) [reply bluepysnap/simulation.py]: Done

File: bluepysnap/simulation.py (hunk 3 above)
Comment (joni-herttuainen) [reply bluepysnap/simulation.py]: Not commenting on the validity of the simulation but, according to the spec, `network` is an optional entry, so I thought it'd be better to not require it being there for the node sets, either.

-> Added a comment

File: bluepysnap/simulation.py (hunk 4 above)
Comment (joni-herttuainen) [reply bluepysnap/simulation.py]: Unless there's another use case for copying a node set, I'm not sure it's worth it.

File: bluepysnap/node_sets.py (hunk 1 above)
Comment (joni-herttuainen) [reply bluepysnap/node_sets.py]: Agreed -> done

File: bluepysnap/node_sets.py [hunk 5]
```diff
@@ -67,7 +68,7 @@ def __init__(self, content, instance):
         Returns:
```
[the remark this answers is not in the document]
Comment (joni-herttuainen) [reply bluepysnap/node_sets.py:69]: -> done

**Review by GianlucaFicarelli — APPROVED** [review]
LGTM

**Pull request merged**